### PR TITLE
Postgres Source Fixed JSONB array mapping

### PR DESCRIPTION
**Pull request by VitaliiMaltsev**
## What
Issue https://github.com/airbytehq/oncall/issues/1570

## How
*Describe the solution*

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

### CODE REVIEW

**Review by subodh1810 — CHANGES_REQUESTED** [review]
Can you add tests for this

File: airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSourceOperations.java [hunk 1]
```diff
@@ -216,7 +216,7 @@ private void putJsonbArray(ObjectNode node, String columnName, ResultSet resultS
     final ResultSet arrayResultSet = resultSet.getArray(colIndex).getResultSet();
 
     while (arrayResultSet.next()) {
-      final PGobject object = getObject(arrayResultSet, colIndex, PGobject.class);
+      final PGobject object = getObject(arrayResultSet, 2, PGobject.class);
```
[the remark this answers is not in the document]
Comment (subodh1810) [reply airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSourceOperations.java:219]: I think you need to traverse the result set and increase the index as you go. This seems wrong

[the remark this answers is not in the document]
Comment (VitaliiMaltsev) [reply airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSourceOperations.java:219]: > Am not sure I understand this. Why is the value hardcoded at 2?

Please take a look how all of other arrays mapping implemented (varchar array, int array, etc)
During iteration over arrays resultSet there are only 2 indexes: 
index 1 - the ordinal number of the element in the array
index 2 - value

[the remark this answers is not in the document]
Comment (subodh1810) [reply airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSourceOperations.java:219]: I see. Can you add a test for this.

[the remark this answers is not in the document]
Comment (VitaliiMaltsev) [reply airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSourceOperations.java:219]: > I see. Can you add a test for this.

we already have test for jsonb[] datatype in AbstractPostgresSourceDatatypeTest. Updated it to contain more than 2 elements in array

**Review by subodh1810 — APPROVED** [review]

**Comment by subodh1810** [comment]
@VitaliiMaltsev we want you to carry out a testing with your fix PR.
1. Run a sync for a Postgres table with jsonb data type to a bigquery destination on version 1.0.49
2. Update the connector version which has your fix 
3. Run the sync again and see if the data written in the destination is correct or not. 

Please share how the data looks in destination on version 1.0.49 and with your fix

**Comment by subodh1810** [comment]
Also another of set @VitaliiMaltsev we want you to carry out a testing with your fix PR.
1. Run a sync for a Postgres table with jsonb data type to a bigquery destination on version 1.0.49
2. Update the connector version which has your fix 
3. Refresh the schema
4. Run the sync again and see if the data written in the destination is correct or not. 

Please share how the data looks in destination on version 1.0.49 and with your fix

**Comment by VitaliiMaltsev** [comment]
> @VitaliiMaltsev tests to carry out with this PR.
> 
> 1. Run a sync for a Postgres table with jsonb data type to a bigquery destination on (postgres version 1.0.49)
> 2. Update the postgres connector version which has your fix
> 3. Run the sync again and see if the data written in the destination is correct or not.
> 
> Please share how the data looks in destination on version 1.0.49 and with your fix

@subodh1810 @edgao 
##Test results

**1.0.49 sync**

**_airbyte_raw_table**



**Normalized table**

1049schema
1049final
--------------------------------------------------------------------------

**1.0.51 sync without schema refresh**

_airbyte_raw_table



**Normalized table**

1049schema
1050final
--------------------------------------------------------------------------

**1.0.51 sync with schema refresh**

**_airbyte_raw_table**



**Normalized table**
1049schema
1050schema-refresh

**Comment by subodh1810** [comment]
@VitaliiMaltsev the normalized table for 1.0.51 without schema refresh doesnt show the data

**Comment by subodh1810** [comment]
@edgao shouldnt the datatype of normalized column after refresh be json instead of string (i am guessing we support json data type in bigquery)

**Comment by subodh1810** [comment]
@VitaliiMaltsev the screenshots are difficult to follow. Can you use order by in your queries so that all the screenshots have the same ordering.

**Comment by VitaliiMaltsev** [comment]
> @VitaliiMaltsev the normalized table for 1.0.51 without schema refresh doesnt show the data, its all null. why is that?

@subodh1810 because actual json schema changed to `"test_column":{"type":"object","oneOf":[{"type":"array"},{"type":"object"},{"type":"number"},{"type":"string"},{"type":"boolean"}],"airbyte_type":"json"}}} `and put all the values to the proper json nodes (see _airbyte_raw table) but normalization used the old schema `"test_column":{"type":"string"}` without schema refresh

**Comment by VitaliiMaltsev** [comment]
> @VitaliiMaltsev the screenshots are difficult to follow. Can you use order by in your queries so that all the screenshots have the same ordering.

Screenshots have the same order
1. Postgres version
2. _airbyte_raw table
3. Normalized table schema
4. Normalized table values

**Comment by subodh1810** [comment]
https://github.com/airbytehq/airbyte/pull/23608#issuecomment-1450604053
I meant the screenshots showing the data.

**Comment by subodh1810** [comment]
also I dont see the values from postgres. Can you share the table structure in postgres (`\d <table_name>`) and how the actual data looks in postgres

**Comment by VitaliiMaltsev** [comment]
> also I dont see the values from postgres. Can you share the table structure in postgres (`\d <table_name>`) and how the actual data looks in postgres

Original Postgres Table
postgres-original-json-table

**Comment by subodh1810** [comment]
@VitaliiMaltsev 
 Update from a quick discussion that I and 
@edgao had. The fact that the version with your changes is writing null values for jsonb columns in the destination if customers dont trigger a refresh schema (as mentioned [here](https://github.com/airbytehq/airbyte/pull/23608#issuecomment-1450543931)) is bad. We decided that the right way to go about this is to revert the original [PR](https://github.com/airbytehq/oncall/issues/1570) and publish new version so that OSS users have a fix and release of postgres is unblocked.

Once done we need to revisit the decision making on how we want to introduce jsonb change and after thorough testing with proper backward compatibility test you can again raise a new PR.

We will have to delete the image 1.0.51 published [here](https://github.com/airbytehq/airbyte/pull/23608#issuecomment-1450294172) and [here](https://github.com/airbytehq/airbyte/pull/23608#issuecomment-1450286361)

**Comment by subodh1810** [comment]
Update : Deleted the 1.0.51 tags from docker hub

**Pull request closed without merging**